### PR TITLE
SNOW-607796: Escape query tag in SQL when it has special characters

### PR DESCRIPTION
**Pull request by sfc-gh-jdu**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-607796. See another fix for `use <object>` in #380. 

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: tests/integ/scala/test_query_tag_suite.py [hunk 1]
```diff
@@ -14,9 +14,18 @@
...
 
-def test_set_query_tag(session):
+@pytest.mark.parametrize(
+    "query_tag",
+    [
+        Utils.random_name_for_temp_object(TempObjectType.QUERY_TAG),
+        "'test_tag'",
+        "te\\st_tag",
+        "te\nst_tag",
+        r"\utest_tag",
```
[the remark this answers is not in the document]
Comment (sfc-gh-kwagner) [reply tests/integ/scala/test_query_tag_suite.py:24]: Can we add more tests for more special characters like spaces?

[the remark this answers is not in the document]
Comment (sfc-gh-jdu) [reply tests/integ/scala/test_query_tag_suite.py:24]: done

**Review by sfc-gh-sfan — APPROVED** [review]

File: src/snowflake/snowpark/session.py [hunk 2]
```diff
@@ -792,7 +793,9 @@ def query_tag(self) -> Optional[str]:
     @query_tag.setter
     def query_tag(self, tag: str) -> None:
         if tag:
-            self._conn.run_query(f"alter session set query_tag = '{tag}'")
+            self._conn.run_query(
+                f"alter session set query_tag = {QUESTION_MARK}", params=(tag,)
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/session.py:797]: Nit: A straight `?` is fine here and even more readable.

**Review by sfc-gh-yixie — APPROVED** [review]

File: src/snowflake/snowpark/session.py (hunk 2 above)
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/session.py:797]: `QUESTION_MARK` already exists and is used by batch insertion.

**Pull request merged**